ag-master/main.py

```diff
@@ -0,0 +1,447 @@
+        abort(403)
+
+    return wrapped
+
+
+def admin_only(func):
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        token = request.json.get("access_token", None)
+        course = request.json.get("course", "cs61a")
+        if (token and is_admin_token(access_token=token, course=course)) or (
+            is_staff(course=course)
+            and is_admin(email=get_user()["email"], course=course)
+        ):
+            semester = request.json.get("semester", "sp21")
+            crs = Course.query.filter_by(name=course, semester=semester).first()
+            if crs:
+                return func(crs, *args, **kwargs)
+            abort(404)
+        return login()
+
+    return wrapped
+
+
+def superadmin_only(func):
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        token = request.json.get("access_token", None)
+        if (token and is_admin_token(access_token=token, course="cs61a")) or (
+            is_staff(course="cs61a")
+            and is_admin(email=get_user()["email"], course="cs61a")
+        ):
+            return func(*args, **kwargs)
+        return login()
+
+    return wrapped
+
+
+########################################
+## Per-course Endpoints
+########################################
+
+
+@app.route("/upload_zip", methods=["POST"])
+@admin_only
+def upload_zip(course):
+    data = request.json
+    file = base64.b64decode(data.get("upload", "").encode("ascii"))
+    name = data.get("filename")
+
+    bucket = storage.Client().get_bucket(BUCKET)
+    blob = bucket.blob(f"zips/{course.name}-{course.semester}/{name}")
+    blob.upload_from_string(file, content_type="application/zip")
+    return dict(success=True)
+
+
+@app.route("/create_assignment", methods=["POST"])
+@admin_only
+def create_assignment(course):
+    data = request.json
+    name = data["name"]
+    file = data["filename"]
+    command = data["command"]
+
+    existing = Assignment.query.filter_by(name=name, course=course.secret).first()
+    if existing:
+        existing.file = file
+        existing.command = command
+        db.session.commit()
+        return dict(assign_id=existing.ag_key)
+
+    id = gen_salt(24)
+    existing = Assignment(
+        name=name, course=course.secret, file=file, command=command, ag_key=id
+    )
+    db.session.add(existing)
+    db.session.commit()
+
+    return dict(assign_id=id)
+
+
+@app.route("/jobs/<job>")
+@admin_only
+def job_info(course, job):
+    job = Job.query.filter_by(job_key=job).first()
+    if not job:
+        abort(404)
+
+    assignment = Assignment.query.filter_by(ag_key=job.assignment).first()
+    if not assignment or assignment.course != course.secret:
+        abort(403)
+
+    return {
+        "assignment": assignment.name,
+        "backup": job.backup,
+        "status": job.status,
+        "result": job.result,
+    }
+
+
+@app.route("/course_info")
+@admin_only
+def course_info(course):
+    assignments = Assignment.query.filter_by(course=course.secret)
```

Comment:
don't you need to `.all()` here? It 500s when I visit the page. Also might be nice to make this the index page